src/ledger-bridge-keyring.ts

```diff
@@ -0,0 +1,908 @@
+): payload is IFrameMessageSignResponsePayload {
+  return typeof (payload as IFrameMessageSignResponsePayload).v === 'string';
+}
+
+function isGetAddressMessageResponse(
+  payload: IFrameMessageResponsePayload,
+): payload is IFrameMessageGetAddressResponsePayload {
+  return (
+    typeof (payload as IFrameMessageGetAddressResponsePayload).publicKey ===
+    'string'
+  );
+}
+
+function isConnectionChangedResponse(
+  payload: IFrameMessageResponsePayload,
+): payload is IFrameMessageConnectionResponsePayload {
+  return (
+    typeof (payload as IFrameMessageConnectionResponsePayload).connected ===
+    'boolean'
+  );
+}
+
+export class LedgerBridgeKeyring extends EventEmitter {
+  static type: string = keyringType;
+
+  readonly type: string = keyringType;
+
+  page = 0;
+
+  perPage = 5;
+
+  unlockedAccount = 0;
+
+  accounts: readonly string[] = [];
+
+  accountDetails: Record<string, AccountDetails> = {};
+
+  hdk = new HDKey();
+
+  hdPath = hdPathString;
+
+  paths: Record<string, number> = {};
+
+  network = 'mainnet';
+
+  implementFullBIP44 = false;
+
+  iframeLoaded = false;
+
+  isDeviceConnected = false;
+
+  currentMessageId = 0;
+
+  _eventListener?: (params: {
+    origin: string;
+    data: IFrameMessageResponse;
+  }) => void;
+
+  messageCallbacks: Record<number, (response: IFrameMessageResponse) => void> =
+    {};
+
+  bridgeUrl: string = BRIDGE_URL;
+
+  iframe?: HTMLIFrameElement;
+
+  delayedPromise?: {
+    resolve: (value: unknown) => void;
+    reject: (error: unknown) => void;
+    transportType: string;
+  };
+
+  constructor(opts = {}) {
+    super();
+    this.deserialize(opts);
+
+    this._setupIframe();
+
+    this._setupListener();
+  }
+
+  serialize() {
+    return Promise.resolve({
+      hdPath: this.hdPath,
+      accounts: this.accounts,
+      accountDetails: this.accountDetails,
+      bridgeUrl: this.bridgeUrl,
+      implementFullBIP44: false,
+    });
+  }
+
+  deserialize(opts: LedgerBridgeKeyringOptions = {}) {
+    this.hdPath = opts.hdPath || hdPathString;
+    this.bridgeUrl = opts.bridgeUrl || BRIDGE_URL;
+    this.accounts = opts.accounts || [];
+    this.accountDetails = opts.accountDetails || {};
+    if (!opts.accountDetails) {
+      this._migrateAccountDetails(opts);
+    }
+
+    this.implementFullBIP44 = opts.implementFullBIP44 || false;
+
+    // Remove accounts that don't have corresponding account details
+    this.accounts = this.accounts.filter((account) =>
+      Object.keys(this.accountDetails).includes(
+        ethUtil.toChecksumAddress(account),
+      ),
+    );
+
+    return Promise.resolve();
+  }
+
+  _migrateAccountDetails(opts: LedgerBridgeKeyringOptions) {
+    if (this._isLedgerLiveHdPath() && opts.accountIndexes) {
+      for (const account of Object.keys(opts.accountIndexes)) {
+        this.accountDetails[account] = {
+          bip44: true,
+          hdPath: this._getPathForIndex(opts.accountIndexes[account] as number),
+        };
+      }
+    }
```

Comment:
To avoid casting `as number` here, we can introduce this non-breaking refactor:
```ts
    if (this._isLedgerLiveHdPath() && opts.accountIndexes) {
      for (const [account, index] of Object.entries(opts.accountIndexes)) {
        this.accountDetails[account] = {
          bip44: true,
          hdPath: this._getPathForIndex(index),
        };
      }
    }
```